[HLSL][DXIL] Implement WaveGetLaneIndex Intrinsic

inbelic opened the pull request:
- add additional lowering for directx backend in CGBuiltin.cpp
    - add directx intrinsic to IntrinsicsDirectX.td
    - add semantic check of arguments in SemaHLSL.cpp
    - add mapping to DXIL op in DXIL.td

    - add testing of semantics in WaveGetLaneIndex-errors.hlsl
    - add testing of dxil lowering in WaveGetLaneIndex.ll
  
Resolves #70105

inbelic committed:
[HLSL][DXIL] Implement WaveGetLaneIndex

    - add additional lowering for directx backend in CGBuiltin.cpp
    - add directx intrinsic to IntrinscsDirectX.td
    - add semantic check of arguments in SemaHLSL.cpp
    - add mapping to DXIL op in DXIL.td

    - add testing of semantics in WaveGetLaneIndex-errors.hlsl
    - add testing of dxil lowering in WaveGetLaneIndex.ll

inbelic committed:
add comment

pow2clk commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18827,9 +18827,21 @@ case Builtin::BI__builtin_hlsl_elementwise_isinf: {
-        {}, false, true));
+    // Since we don't define a SPIR-V intrinsic for the SPIR-V built-in from
+    // SPIRVBuiltins.td, manually get the matching name for the DirectX
+    // intrinsic and the demangled builtin name
```
I fear I missed some discussion explaining this, but why don't we add a builtin to SPIRVBuiltins.td?

pow2clk commented on clang/test/SemaHLSL/BuiltIns/WaveGetLaneIndex-errors.hlsl:
```diff
@@ -0,0 +1,6 @@
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -emit-llvm-only -disable-llvm-passes -verify -verify-ignore-unexpected
```
Are you getting unexpected diagnostics? If so, I'd like to understand them. If not, I don't think `-verify-ignore-unexpected` is needed. If anything, we should limit it to the type of diagnostic that is being produced e.g. `-verify-ignore-unexpected=note`

inbelic replied:
The flag was not needed, so have removed it.

pow2clk reviewed (commented):
Small typo in description: 
"IntrinscsDirectX"

Looks good, but I have a question about the SPIRV part.

inbelic committed:
review comments:

    - remove flags that are not needed
    - update intrinsic naming to wave_getlaneindex to follow llvm conventions

pow2clk reviewed (approved):
Looks good! I since remembered that I can edit descriptions and fix small typos like that in the future 🙃
